### DISCUSSION TITLE
Ideas for v3

### DISCUSSION THREAD

**Discussion by dai-shi** [pmndrs/jotai#2889] [Ideas]
Nothing determined. ETA is year 2026.

### Plan for v3

- Drop old TS versions
- Drop UMD/SystemJS builds
- Drop old Node versions
- Module-first package.json
- Drop CJS too
- Use `process.env.NODE_ENV` directly in the code
- Drop rollup in favor of tsc
- Drop React < 18
- Move `atomFamily` util to `jotai-family` package (deprecate it during v2)
- Drop `loadable` util in favor of `unwrap` (deprecate it during v2)
- Move `jotai/babel` to `jotai-babel` package (deprecate it during v2)
- Drop `setSelf` option (deprecate it during v2)
- Do not use `use` in useAtomValue 👉  https://github.com/pmndrs/jotai/pull/3133#issuecomment-3212629114
- Redesign buildingBlocks array structure

### Not for v3

- Consider dropping `INTERNAL_overrideCreateStore` if jotai-devtools can have a good DX without it.
- Consider dropping `.onMount` if `withMount` util works in v2
- Consider dropping `signal` option if `withSignal` util works in v2

**Comment by ryansobol** [comment]
Perhaps consider adding aliases for the `useAtomValue` and `useSetAtom` functions that more closely align to the concept of `read` and `write` atom config functions. For example:

 
- `useAtomValue` aliased to `useAtomRead`
- `useSetAtom` aliased to `useAtomWrite`

Benefits:

- Would ease the onboarding for newcomers trying to construct a mental model for atom config functions
- Would ease code reviewing for all trying to understand the expansion/contraction of component behavior to/from the `useAtom` function 
- Would allow for experimenting with new names without breaking existing code, deprecating and providing a migrator tool when appropriate

**Reply by dai-shi** [reply]
Naming is hard. There will be pros and cons. At a glance, `useAtomRead` is confusing to me as I would expect it returns the `read` function. Thanks for the suggestion, but I don't believe we should change this.

**Comment by dmaskasky** [comment]
How about:
- migrating all `atomWith*` utilities in jotai/utils to `with*`?
- moving dev store methods to `createDevStore` under jotai-utils

**Reply by dai-shi** [reply]
If I were to take those ideas, my plan would be to do these during v2,
- providing `with*` utils by `jotai-utils` package
- moving `deriveDevStore` to `jotai-devtools` package

Note that major version bump of libraries is nothing fancy. It's best if people wouldn't notice anything with migrating.

**Reply by dmaskasky** [reply]
If we decide to do both, I think we should probably make both available in v2 and deprecate both original apis. Then remove them completely in V3. It would be less api for the user to grok and less for us to maintain.

**Reply by sebinsua** [reply]
@dmaskasky Regarding `atomWith*` to `with*` is that purely a naming change or is the idea for these functions to receive atoms and return atoms so they are kind of high-ordered wrappers?

**Reply by dmaskasky** [reply]
> @dmaskasky Regarding `atomWith*` to `with*` is that purely a naming change or is the idea for these functions to receive atoms and return atoms so they are kind of high-ordered wrappers?

The latter. So you can do
```ts
const countAtom = atom(0)
comst countWithLoggerAndStorage = withLogger(withStorage(countAtom))
```

**Comment by a10nik** [comment]
What about reducing the coupling with react? It's an optional dep already, what about taking a step further and extracting the react-specific logic into a separate package? It might pave a path for the other framework integrations to follow. BTW, are there any plans to start advertising jotai as a multi-framework solution? Are there any react/Suspense-specific hacks in the core logic now?

I've seen Vue integration discussed in #1811, however the 3rd-party package mentioned there [jotai-vue](https://github.com/wobsoriano/jotai-vue) seems dead

**Reply by dai-shi** [reply]
> Are there any react/Suspense-specific hacks in the core logic now?

Not particularly (except for one I would like to refactor). But I don't have any such plans either. I'd say it's more about our React-first mental model for Zustand, Jotai, and Valtio. (So, it's not just about Jotai.)

Unless there were a huge demand (like if Astro said they would pick it as their official state management solution), I wouldn't go that direction.

> the 3rd-party package mentioned there [jotai-vue](https://github.com/wobsoriano/jotai-vue) seems dead

That actually tells the demand.

**Reply by dai-shi** [reply]
I don't mean I never thought about it. I created https://www.npmjs.com/package/jotai-core for that purpose.

What I mean is the decision wouldn't come from looking at React users. So, my question is if there will be dedicated long-term contributors and users for jotai-vue, jotai-svelte and so on.

**Comment by marcebdev** [comment]
I'm not sure on feasibility but being able to have some `.peek` like functionality (not that naming has to be this, but to have an example) maybe even something like

```
const peekForm = usePeekAtom(myFormAtom)

function formSubmit() {
  const formState = peekForm()
  ...
}
```

When a value can be explicitly checked on a specific event call (like form state on onSubmit) it would be really nice & beneficial because you wouldn't have to subscribe the parent to the state at all.

**Reply by dai-shi** [reply]
It should be already possible.

```js
const usePeekAtom = (atom) => {
  const store = useStore();
  return () => store.get(atom);
};
```

**Reply by dmaskasky** [reply]
You can literally do anything with the new Jotai API. 🦄

```ts
type GetterWithPeek = Getter & { peek: Getter }

function atomWithPeak<T>(
  read: (getterWithPeek: GetterWithPeek, options: Options) => T
) {
  const refAtom = atom<{ peek: GetterWithPeek }>(() => ({}))
  refAtom.debugPrivate = true
  const baseAtom = atom<T>((get, options) => {
    const { peek } = get(refAtom)
    const getterWithPeek = Object.assign(get, { peek })
    return read(getterWithPeek, options)
  })
  baseAtom.onInit = (store) => {
    store.get(refAtom).peek = store.get
  }
  return baseAtom
}
```

**Reply by dai-shi** [reply]
I don't think the refAtom hack is necessary in this case. Or, I'm not sure if it triggers extra renders.

I would also avoid modifying `get`. Just a preference though.
```js
function atomWithPeek(read) {
  const peekAtom = atom();
  const baseAtom = atom((get, options) => {
    const getterWithPeek = (a) => get(a);
    getterWithPeek.peek = get(peekAtom);
    return read(getterWithPeek, options);
  };
  baseAtom.unstable_onInit = (store) => {
    const peek = (a) => store.get(a);
    store.set(peekAtom, peek);
  };
  return baseAtom;
}
```

**Reply by dmaskasky** [reply]
Looks good.

I don't think enough people are aware of the power of this new api. It might be a good candidate for a blog post or a tweet maybe.

**Reply by dai-shi** [reply]
It's too powerful and not recommended for most app developers, I'd rather think.

**Reply by marcebdev** [reply]
That's really cool, thank you so much for sharing!

**Reply by marcebdev** [reply]
@dai-shi would you happen to have an example of how the `atomWithPeek` util would be used? I'm trying to give it a shot and realize I don't really understand how the `read` argument works, or at least how I end up being able to peek the actually works. Ended up trying

```
const testAtom = atom('hello world')
const peekableAtom = atomWithPeek((get) => get(testAtom))

...

const peek = useAtomValue(peekableAtom)
```

I think I don't have the types correct atm but I got to something like this. I couldn't use the `Getter` type directly bc it would complain that `.peek` doesn't exists on `getterWithPeek.peek` 
```
export function atomWithPeek<T>(read: Read<T>) {
  const peekAtom = atom<PeekAtom<T>>()

  const baseAtom = atom((get, options) => {
    const getterWithPeek = <Value>(a: Atom<Value>) => get(a)
    getterWithPeek.peek = get(peekAtom)
    return read(getterWithPeek, options)
  })

  baseAtom.unstable_onInit = (store) => {
    const peek = <Value>(a: Atom<Value>) => store.get(a)
    store.set(peekAtom, peek)
  }

  return baseAtom
}
```

**Reply by dmaskasky** [reply]
https://codesandbox.io/p/devbox/unruffled-taussig-yj7q5f?workspaceId=ws_C9Ksjj78sgjEf6XMmWZ1Fr

**Reply by dai-shi** [reply]
@marcebdev FWIW, I'd recommend usePeekAtom as a straightforward solution.
https://github.com/pmndrs/jotai/discussions/2889#discussioncomment-12834959

**Reply by marcebdev** [reply]
Thank y'all both so much! Honestly I didn't even realize `useStore` just works with the Provider I thought you could only use it with `createStore` so I just tried it and that worked well and was super simple too!

**Reply by organize** [reply]
I've found `useAtomCallback` helpful when I want "atom scope" non-reactively:

```
const formSubmit = useAtomCallback(
 useCallback((get, _set) => {
  const form = get(myFormAtom)
  ...
 }, [])
)
```

It's particularly useful when dealing with a lot of dependencies and requiring setters. It also allows me to arbitrarily batch `set` calls (as long as you don't also `get` the atom you're setting).

**Comment by himself65** [comment]
> Do not use use in useAtomValue (migration path: add { use: true } option which is true by default in v2, and warns it)

If remove the `use`, I think it would looks weird?

```
const value = use(useAtomValue(atom))
```

**Reply by dai-shi** [reply]
Yeah, but I don't think it's too weird. It's not yet widely used, but `use` is still new.

**Comment by dai-shi** [comment]
https://react.dev/blog/2025/04/23/react-labs-view-transitions-activity-and-more#concurrent-stores
I'll wait for it...

**Reply by dmaskasky** [reply]
I read the page but I still don't understand what has been planned in terms of functionality.

**Comment by a10nik** [comment]
> Drop React < 18
> Do not use use in useAtomValue

BTW `use()` is only available since react v19. Folks with v18 should be fine as long as `{ use: true }` exists but once it's obsolete we should either drop v18 as well or recommend a use() polyfill, right? Likely, this one? https://github.com/dai-shi/react18-use

**Reply by dai-shi** [reply]
Yes, that's what I did in [valtio](https://github.com/pmndrs/valtio/blob/ccc47e4ab50630aeb0321bd0302498f36158d9ab/README.md?plain=1#L170).

**Comment by organize** [comment]
For `useAtomValue`, I find myself wanting for a couple of things:

1. The ability to pass in a custom equality check in place of `Object.is`
2. The ability to pass in an ad-hoc selector (e.g. `(value: Value) => Out`)

The first one seems simple implementation-wise, but I'm not sure of the implications. Use cases are a bit niche, but it might be reasonable band-aid in some cases.

Second one less so. Basically, I just want to avoid re-renders, but having to write tons of `selectAtom`/`focusAtom` is slightly tedious, and while I can write a home-brew hook for this, it _should_ be a fairly common use case.

Any thoughts? Does this make any sense?

**Reply by dai-shi** [reply]
I assume both are niche, or not typical. I'd suggest creating your home-brew custom hook. Or, if you would like to create a third-party library, it'd be an option.

(The rationale is that Jotai is developed to avoid such selectors (or custom equality function) with atoms.)

**Comment by iwoplaza** [comment]
> Consider dropping unwrap util (migration path: recommend loadable or jotai-derive)

Currently, both `unwrap` and `loadable` use their own internal cache for tracking the status of Promises, which means if we want to inspect if an async atom has already fulfilled or not, we always need to wait a tick before getting its value. I wanted to implement a way to inspect eager atoms in this way for `jotai-derive` (now `jotai-eager`), but that would essentially be duplicating what `unwrap` and `loadable` already do, just with a shared Promise cache.

For now, I reimplemented `unwrap` and `loadable` in https://github.com/jotaijs/jotai-eager/pull/13, with a few changes:
- Shared Promise cache, so they can share information about their statuses and reduce unnecessary suspensions.
- Renamed `unwrap` to `withPending`, with the ability to provide a pending fallback (just like `unwrap`) that can be based on the previous value, the pending promise, or the value of another atom.
- Removed memoization from `withPending`. This is because it wasn't obvious that the second argument to `unwrap` has to be stable in order for it to not return a new atom on every call. The name change to `withPending` was also motivated by trying to discourage devs from running it in hot paths, as it's understood that `with*` calls create a new atom.

Since jotai-eager's `loadable` has the same behavior as jotai's `loadable` for vanilla atoms, it would be possible to drop both `loadable` and `unwrap` utils in Jotai v3.

**Reply by dai-shi** [reply]
Yeah, that's one idea. btw, that reminds me of using `attachPromiseStatus` instead of WeakMaps.

**Comment by wojtekmaj** [comment]
Regarding dropping CJS - I don't think this should even be a question. 99,9999% of Jotai users are using some kind of bundler, all of which support ESM just fine. The remaining 0,0001% would be users using pure ESM modules, in which case... Well... They're using ESM modules :D

**Reply by dai-shi** [reply]
Quite a few users may use Jotai in SSR, and they use it without knowing it. Recent Node.js versions are capable of importing ESM from CJS, so it will be fine.

**Comment by wojtekmaj** [comment]
I'm concerned about dropping `setSelf`. I have a complex set of atoms that work together to create kind of "atomWithStorageWhileRevalidate" - optimistically loading user data from localStorage while at the same time loading the latest data from the server. Without `setSelf`, I don't see the possibility of it working.

My code:

```ts
const defaultCurrentUserState = atomWithDefault<User | null | Promise<User | null>>(
  (_get, { setSelf }) => {
    async function getUser(): Promise<User | null> {
      const response = await (networkGet('/api/users/current').catch(
        () => null,
      ) as Promise<User | null>);
      localStorage.setItem('user', JSON.stringify(response));
      return response;
    }

    const rawLocalStorageValue = localStorage.getItem('user');

    if (rawLocalStorageValue) {
      // Update the user in the background while we return the cached value
      getUser().then((nextValue) => {
        if (!isEqual(nextValue, rawLocalStorageValue)) {
          setSelf(nextValue);
        }
      });

      try {
        return JSON.parse(rawLocalStorageValue) as User | null;
      } catch {}
    }

    return getUser();
  },
);

export const currentUserState = atom<
  User | null | Promise<User | null>,
  [SetStateAction<User | null>],
  void
>(
  (get) => get(defaultCurrentUserState),
  async (get, set, nextValueOrSetter: SetStateAction<User | null>) => {
    const nextValue =
      typeof nextValueOrSetter === 'function'
        ? nextValueOrSetter(await get(defaultCurrentUserState))
        : nextValueOrSetter;

    localStorage.setItem('user', JSON.stringify(nextValue));
    set(defaultCurrentUserState, nextValue);
  },
);

**Reply by dmaskasky** [reply]
I was going to recommend moving the side-effect to onMount or to an atomEffect, but I noticed you have the side-effect on every read. Since overwritten atom is set after its read, its marked as dirty and recomputeDependents is causing an additional network request.

When reading the atom when the server has a different value:
------------
1. atom read
1. network request
1. atom write (since not equal)
1. atom read (because recomputeDependents)
1. another network request <---- **EXTRA**

I don't think this is the behavior you want, particularly that your state update is triggered by atom read, which will not trigger again even if the backend state has changed. This is because the value is cached and atomRead is only recomputed when dependents have changed.

But if this code works for you then perhaps this edge case isn't important. If that is the case, either using onMount or atomEffect would be a good fit. Here's atomEffect reimplementation.

```ts

import { atom } from 'jotai'
import { atomWithDefault } from 'jotai/utils'
import { withAtomEffect } from 'jotai-effect'

async function fetchCurrentUser(signal?: AbortSignal): Promise<User | null> {
  const response = await (networkGet('/api/users/current', { signal }).catch(
    () => null,
  ) as Promise<User | null>)

  localStorage.setItem('user', JSON.stringify(response))
  return response
}

function readCachedUser(): User | null | undefined {
  const raw = localStorage.getItem('user')
  if (!raw) return undefined
  try {
    return JSON.parse(raw) as User | null
  } catch {
    return undefined
  }
}

const defaultCurrentUserBaseAtom = atomWithDefault<
  User | null | Promise<User | null>
>(() => {
  const cached = readCachedUser()
  if (cached !== undefined) return cached
  return fetchCurrentUser()
})

const defaultCurrentUserState = withAtomEffect(
  defaultCurrentUserBaseAtom,
  (_get, set) => {
    const cached = readCachedUser()
    if (cached === undefined) return

    const controller = new AbortController()
    const { signal } = controller

    ;(async () => {
      const nextValue = await fetchCurrentUser(signal)
      if (signal.aborted) return
      if (!isEqual(nextValue, cached)) {
        set(defaultCurrentUserBaseAtom, nextValue)
      }
    })()

    return () => {
      controller.abort()
    }
  },
) as typeof defaultCurrentUserBaseAtom

export const currentUserState = atom<
  User | null | Promise<User | null>,
  [SetStateAction<User | null>],
  void
>(
  (get) => get(defaultCurrentUserState),
  async (get, set, nextValueOrSetter: SetStateAction<User | null>) => {
    const nextValue =
      typeof nextValueOrSetter === 'function'
        ? nextValueOrSetter(await get(defaultCurrentUserState))
        : nextValueOrSetter

    localStorage.setItem('user', JSON.stringify(nextValue))
    set(defaultCurrentUserState, nextValue)
  },
)
```

**Reply by wojtekmaj** [reply]
Wow, this DOES work perfectly! Thank you SO much! I absolutely did not expect that from you 😅

**Comment by nsimonson** [comment]
One thing I'd like to consider for v3 is adding lifecycle control for the providerless default store.

We have a large application that intentionally uses Jotai's default store both through providerless React hooks and imperatively through `getDefaultStore()`.

In retrospect, we could have created our own singleton with `createStore()` and consistently passed it through `<Provider>`. That would have given us explicit control over the store lifecycle. But `getDefaultStore()` is a public API, and the providerless approach was a natural choice for sharing state between React and non-React code.

That pattern is now deeply entrenched. We have 400+ modules that obtain the default store, including modules that capture it at initialization:

```ts
const store = getDefaultStore()
```

The problem is test isolation.

Tests running sequentially in the same Vitest worker share Jotai's module-level default store. There is currently no public way to return that store to a pristine state between tests.

Using a fresh Provider in each test doesn't solve this because non-React application code still accesses the default store. Replacing the default Store object isn't sufficient either because existing modules may already hold references to it.

The only way we've found to guarantee isolation without restructuring a large portion of the application is to reach into Jotai's internals.

We're currently prototyping a test shim using `INTERNAL_overrideCreateStore` and `INTERNAL_buildStoreRev3`. At a high level, it replaces Jotai's WeakMap-backed atom state with clearable Maps while preserving the Store object's identity:

```ts
const resets = new Set<() => void>()

INTERNAL_overrideCreateStore(() => () => {
  const atomState = new Map()
  const invalidatedAtoms = new Map()
  const changedAtoms = new Set()

  resets.add(() => {
    atomState.clear()
    invalidatedAtoms.clear()
    changedAtoms.clear()
  })

  return INTERNAL_buildStoreRev3(
    atomState,
    undefined,
    invalidatedAtoms,
    changedAtoms,
  )
})

export function resetJotaiStores() {
  for (const reset of resets) {
    reset()
  }
}
```

This is deliberately simplified to illustrate the approach. Our actual prototype also accounts for things like mounted state and pending async work.

The important property is that the Store identity does not change. Code that previously did:

```ts
const store = getDefaultStore()
```

keeps the same reference while the state behind it can be reset between tests.

The prototype demonstrates the behavior we need, but depending on `INTERNAL_overrideCreateStore`, `INTERNAL_buildStoreRev3`, and Jotai's internal building-block layout isn't something we'd want to rely on long term.

What seems to be missing from the public API is a lifecycle boundary for the default store.

Ideally it could be as simple as:

```ts
beforeEach(() => {
  resetDefaultStore()
})
```

After that call:

- providerless React should observe fresh atom state
- imperative code using `getDefaultStore()` should observe fresh atom state
- previously obtained references to the default store should remain valid

For our test use case, it wouldn't need to support arbitrary resets while an application is actively mounted.

Jotai's use of WeakMaps internally doesn't seem like it should prevent this. A reset wouldn't clear the existing WeakMaps. That isn't possible. Our Map-backed shim is only necessary because we're implementing the reset externally through Jotai's internal APIs. A first-class implementation could preserve the public Store identity while replacing its backing state with fresh internal containers.

This feels particularly relevant for v3 because solving it may require changing the semantics or implementation of the default store while remaining effectively invisible to applications that don't use the new lifecycle API.

Would this be something you'd consider supporting in v3?

**Reply by dai-shi** [reply]
I wouldn't recommend using the default store in that case. Is it correct that using `createStore()` solves you testing issues? If you don't prefer `<Provider>`, why not create custom hooks?

**Reply by nsimonson** [reply]
I agree that `createStore()` explicitly passed through `<Provider>` is what I would choose if we were starting over. My concern is that the default API makes it very easy to get into this situation without realizing it.

Once `getDefaultStore()` is used imperatively, there is no supported way to establish a fresh default store between tests in the same worker. Moving to an explicitly owned store solves that, but it is a migration away from the architecture Jotai provides by default.

So I think the part I’m trying to understand is whether you consider that an intentional limitation of the default store. Is the expectation that the default store should effectively live for the lifetime of the Jotai module, or do you think providing some lifecycle control over it would be feasible in v3?

**Reply by dai-shi** [reply]
Yes, I consider Provider-less mode (the default store) comes with limitations and for easy use cases.
Improving docs to clarify that would be great.
